### PR TITLE
Dask feature creation

### PR DESCRIPTION
**Pull request by kirilklein**
The most important changes involve the refactoring of feature creation and exclusion methods, the addition of a new data loader class, and updates to various configuration files.

### Refactoring and Code Improvements:

* [`corebehrt/classes/excluder.py`](diffhunk://#diff-edd3297efa0140f6e017c26ba3c4a6f26b9a266ddcf85a747978e7c928a146caL5-L14): Refactored the `Excluder` class to use `dask.dataframe` instead of `pandas` and simplified the `exclude_incorrect_events` and `exclude_short_sequences` methods by removing unnecessary steps and dependencies. [[1]](diffhunk://#diff-edd3297efa0140f6e017c26ba3c4a6f26b9a266ddcf85a747978e7c928a146caL5-L14) [[2]](diffhunk://#diff-edd3297efa0140f6e017c26ba3c4a6f26b9a266ddcf85a747978e7c928a146caL30-R36)
* [`corebehrt/classes/features.py`](diffhunk://#diff-b722d8130331f234335bdc116f71a2d23318d35aa06a0a0d57b398281d7716dfL2-R32): Refactored the `FeatureCreator` class to include detailed docstrings, added new methods for preparing concepts and checking required columns, and updated the `__call__` method to streamline feature creation and exclusion processes. [[1]](diffhunk://#diff-b722d8130331f234335bdc116f71a2d23318d35aa06a0a0d57b398281d7716dfL2-R32) [[2]](diffhunk://#diff-b722d8130331f234335bdc116f71a2d23318d35aa06a0a0d57b398281d7716dfL28-R100)

### New Functionality:

* [`corebehrt/classes/loader.py`](diffhunk://#diff-012462f80bf201c72551037e98cd75717df52cd0b05687aa8efa1762f4fd5b39R1-R72): Introduced the `FormattedDataLoader` class to load concepts and patient information from a specified folder, ensuring the presence of required columns and handling multiple concept types.

### Configuration Updates:

* [`corebehrt/configs/create_data.yaml`](diffhunk://#diff-ab8cde67732406d236558a0dffd677e74796244fa4130fb69a9e48c342376943L9-L15): Updated the configuration to include `features_dir` and revised the `concept_types` list.
* [`corebehrt/configs/finetune.yaml`](diffhunk://#diff-056e5724877ef201a75a1cd4d5a8a0144fde546e6a39efba886c5b0af57b57caL23-R23): Increased the number of patients for fine-tuning and adjusted the censoring time for outcomes. [[1]](diffhunk://#diff-056e5724877ef201a75a1cd4d5a8a0144fde546e6a39efba886c5b0af57b57caL23-R23) [[2]](diffhunk://#diff-056e5724877ef201a75a1cd4d5a8a0144fde546e6a39efba886c5b0af57b57caL39-L42)
* [`corebehrt/configs/pretrain.yaml`](diffhunk://#diff-bb5b3f80259061913b76c14f2d8b64cbfcb48de3a769b1b99822c0c6fe23761bL9-R9): Minor update to the `predefined_pids` path configuration.

### Bug Fixes and Cleanups:

* [`corebehrt/common/setup.py`](diffhunk://#diff-a5010fadad9cfe42518d1b0c6b44497809fc8f212d208aa6cec4baba7a19d617L79): Removed an unnecessary directory creation step in the `prepare_directory` method.
* [`corebehrt/common/utils.py`](diffhunk://#diff-ead1c55245c9cd89f7d6c3df267372186ca98124b2e01d6575e8c12c7969c9b4L41-L52): Removed the `check_directory_for_features` function as it was no longer needed.

### CODE REVIEW

**Comment by kirilklein** [comment]
Many of these 48 changed files are in testing, including updated data
Code related changes (excluding tests) only occur in about 10 files.

File: corebehrt/data/prepare_data.py [hunk 1]
```diff
@@ -188,6 +188,7 @@ def _prepare_mlm_features(self, predefined_splits, val_ratio=0.2) -> Data:
         data_cfg = self.cfg.data
         model_cfg = self.cfg.model
         paths_cfg = self.cfg.paths
+        paths_cfg = self.cfg.paths
```
Comment (Montgomeryyyy) [inline corebehrt/data/prepare_data.py:191]: double?

File: corebehrt/data/prepare_data.py [hunk 2]
```diff
@@ -227,6 +228,7 @@ def _prepare_mlm_features(self, predefined_splits, val_ratio=0.2) -> Data:
         # 5. Truncation
         logger.info(f"Truncating data to {data_cfg.truncation_len} tokens")
         data = truncate_data(data, data_cfg.truncation_len, vocab, truncate_patient)
+        data = truncate_data(data, data_cfg.truncation_len, vocab, truncate_patient)
```
Comment (Montgomeryyyy) [inline corebehrt/data/prepare_data.py:231]: also double

File: corebehrt/functional/utils.py [hunk 3]
```diff
@@ -92,6 +81,8 @@ def get_abspos_from_origin_point(
     timestamps: Union[pd.Series, List[datetime]], origin_point: datetime
 ) -> Union[pd.Series, List[float]]:
     """Get the absolute position in hours from the origin point"""
+    if isinstance(timestamps, dd.Series):  # len can cause issues with dask
+        return (timestamps - origin_point).dt.total_seconds() / 60 / 60
```
Comment (Montgomeryyyy) [inline corebehrt/functional/utils.py:85]: I thought we wanted something less fine-grained than hours?

File: corebehrt/test_create_features.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
+
+import dask.dataframe as dd
```
Comment (Montgomeryyyy) [inline corebehrt/test_create_features.py:2]: is this a test, and if so should it not be in tests?

**Review by Montgomeryyyy — COMMENTED** [review]
I don't have too many comments:)

File: corebehrt/classes/loader.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        ]
+        concepts = dd.concat(concepts)
+        self.check_concepts_columns(concepts)
+
+        patients_info = self.load_patients_info()
+        self.check_patients_info_columns(patients_info)
+
+        return concepts, patients_info
+
+    def check_concepts_columns(self, concepts: dd.DataFrame):
```
[the remark this answers is not in the document]
Comment (kirilklein) [reply corebehrt/classes/loader.py]: Added in f7af9c84e1b1ee58c3dc95e369976167f5aff056

File: corebehrt/functional/utils.py (hunk 3 above)
Comment (kirilklein) [reply corebehrt/functional/utils.py:85]: That will be adressed in a future PR:)

File: corebehrt/data/prepare_data.py (hunk 1 above)
Comment (kirilklein) [reply corebehrt/data/prepare_data.py]: adressed in 81aba9ebe7b955d6f86e36d2783a7c2dae12cb9a

File: corebehrt/test_create_features.py (hunk 4 above)
Comment (kirilklein) [reply corebehrt/test_create_features.py:2]: That wasn't meant to be tracked

**Review by mikkelfo — APPROVED** [review]
Looks fine to me.

Only comment (out of scope for this PR), is that the validation of columns `check_concepts_columns` and `check_patients_info_columns` should probably be done in a more structured way (so we don't have to call it a lot of different places)

**Review by Montgomeryyyy — APPROVED** [review]
Looks good:)

**Pull request merged**